parser/lexer.go

```diff
@@ -516,7 +516,7 @@ func (r *reader) peek() rune {
 	v, w := rune(r.s[r.p.Offset]), 1
 	switch {
 	case v == 0:
-		return unicode.ReplacementChar
+		return v // illegal UTF-8 encoding
```

Comment:
The width `w` need to be set too? Then we don't need to check this case `v == 0'.